editor/modes/visual-editor/block.js

```diff
@@ -330,31 +331,37 @@ class VisualEditorBlock extends Component {
 		/* eslint-disable jsx-a11y/no-static-element-interactions, jsx-a11y/onclick-has-role, jsx-a11y/click-events-have-key-events */
 		return (
 			<div
-				ref={ this.bindBlockNode }
-				onKeyDown={ this.onKeyDown }
-				onFocus={ this.onFocus }
+				ref={ ( ref ) => this.props.blockRef( ref ) }
 				onMouseMove={ this.maybeHover }
 				onMouseEnter={ this.maybeHover }
 				onMouseLeave={ onMouseLeave }
 				className={ wrapperClassname }
 				data-type={ block.name }
-				tabIndex="0"
-				aria-label={ blockLabel }
 				{ ...wrapperProps }
 			>
 				<BlockDropZone index={ order } />
-				{ ( showUI || isHovered ) && <BlockMover uids={ [ block.uid ] } /> }
-				{ ( showUI || isHovered ) && <BlockRightMenu uid={ block.uid } /> }
+				{ ( showUI || isProperlyHovered ) && <BlockMover uids={ [ block.uid ] } /> }
+				{ ( showUI || isProperlyHovered ) && <BlockRightMenu uids={ [ block.uid ] } /> }
 				{ showUI && isValid && mode === 'visual' && <BlockToolbar uid={ block.uid } /> }
-
-				{ isFirstMultiSelected && (
+				{ isFirstMultiSelected && ! this.props.isSelecting &&
 					<BlockMover uids={ multiSelectedBlockUids } />
-				) }
+				}
+				{ isFirstMultiSelected && ! this.props.isSelecting &&
+					<BlockRightMenu
+						uids={ multiSelectedBlockUids }
+						focus={ true }
+					/>
+				}
 				<div
+					ref={ this.bindBlockNode }
 					onKeyPress={ this.maybeStartTyping }
 					onDragStart={ ( event ) => event.preventDefault() }
 					onMouseDown={ this.onPointerDown }
+					onKeyDown={ this.onKeyDown }
+					onFocus={ this.onFocus }
 					className="editor-visual-editor__block-edit"
+					tabIndex="0"
```

Comment:
I'm a bit concerned about this change (moving the tabIndex here). Because we heavily rely on the `.editor-visual-editor__block` being tabbable (for arrow navigation for example). I'll have to test the places we use this className.

Reply:
Hm, it's unfortunate that we rely on `.editor-visual-editor__block`. :/ We do have to move it because nesting buttons in this is causing bugs in a Firefox and Safari.

Reply:
I'll update the selector.

Reply:
Yes, but I don't know honestly what's the best alternative would be. This is all related to focus management which is causing us a lot of trouble in several places. We'll have to rethink it entirely IMO:

 - Clarify the role of the focus config in state
 - Have clear guidelines on when to rely on DOM imperative focus handling and state-based focus updates.
 - All this while ideally hiding everything from the block's author. 

I think we should tackle this clarification as soon as we can.

Reply:
Any ideas on what to do next?

> All this while ideally hiding everything from the block's author.

I'm not sure how it relates to this PR.

Reply:
How can I resolve this PR? Should I remove the bug fixes for Firefox and Safari (also broken in master).

Reply:
It relates more to the Arrow navigation which is causing issue with this PR.

Maybe it's ok to do a temporary hack in this PR, adding a way to identify the block's tabbable element (className or attribute or something else) and update the `WritingFlow` component to target this node instead of `editor-visual-editor__block`

Reply:
Ah, ok. I'll update the temporary hack in `WritingFlow`.

Reply:
Rewritten in dc9859f2935c5add1d4c10edf45ca916542ff4e1.

Reply:
>Yes, but I don't know honestly what's the best alternative would be. This is all related to focus management which is causing us a lot of trouble in several places. We'll have to rethink it entirely IMO

Noting that I have a local work-in-progress branch which takes a stab at this with a few ideas:

- We don't explicitly target the block class, but rather allow descendant DOM nodes of WritingFlow assign a broadly-understood `data-` attribute which defines a focus context (for a block, `data-focus-context={ block.uid }`
- State tracks the ID of the current focus context, and when this changes, WritingFlow shifts focus to the first available input within the new context (via `querySelector`) if the context does not already currently have focus
- Arrow key navigation works simply by updating the focus ID in state using the closest focus context of the next eligible input, if it is different than the current value, otherwise simply moving focus
- Blocks can still explicitly move focus to inputs (like video block to its caption) by assigning a custom focus context ID to editable, so e.g. `<Editable focusId="caption" />`, `setFocus( 'caption' )`

As a work-in-progress, some of these ideas are still muddy, for example directionality of focus and shifting focus within the same focus context might be deserving of more controlled treatment within state.